### PR TITLE
feat(cheatcodes): forge execution context check

### LINKED ISSUE #2900: Ability to detect different "environments" in scripts/tests
### Component

Forge

### Describe the feature you would like

I'm working on a more complex setup that includes a bunch of `ffi` sanity checks and post-deploy scripts. It would be great if I could somehow be able to detect when `--broadcast` is enabled so that I see the current script run as a "dry-run" and not actually overwrite my current deployments.

Other use-cases include disabling a bunch of these sanity checks (which take up a lot of extra time) by passing in some user-argument.

There actually seems to be no easy way to discern different "modes", like (e.g. dry-run, or other user-defined parameters).
Running `DRY_RUN=true && forge script deploy` doesn't let me read `DRY_RUN` through `vm.getEnv`, because it's not in the default `.env` file. The closest I've gotten is to being able to detect `--ffi` through a `try vm.ffi() {} catch {}`, which somewhat works for my use-case for now, by always enforcing it to be set to true when broadcasting.

Some things that would be helpful for me include:
- being able to detect `--broadcast`
- being able to read env vars passed in as `DRY_RUN=true && forge script deploy`
- being able to pass in a different environment `--env .env-test`
- being able to read additional .toml settings (and then running via `FOUNDRY_PROFILE=mainnet-dry-run && forge script deploy`)

I think most helpful and straightforward would be nr2 or nr3.
Just curious if anyone else has found the need for something like this.

### Additional context

Edit:
Decided to add some additional context, because I'm not sure it's really clear what I'm asking for.
Below is my setup, and I'm basically using (abusing) `--ffi` as a way to determine `DRY_RUN=true`, since this has to be enabled anyway to run some of my sanity checks.

```solidity
contract deploy is DeployScripts {
    function isFFIEnabled() internal returns (bool) {
        string[] memory script = new string[](1);
        script[0] = "echo";
        try vm.ffi(script) {
            return true;
        } catch {
            return false;
        }
    }

    function startBroadcastIfFFIEnabled() internal {
        if (isFFIEnabled()) {
            vm.startBroadcast();
        } else {
            console.log('FFI disabled: run again with `--ffi` to save deployments and run storage compatibility checks.');
            console.log('Disabling `broadcast`, continuing as a "dry-run".\n');

            __DEPLOY_SCRIPTS_DRY_RUN = true;

            // need to start prank instead now to be consistent in "dry-run"
            vm.stopBroadcast();
            vm.startPrank(msg.sender);
        }
    }

    function run() external {
        startBroadcastIfFFIEnabled();

        setUpContracts();

        if (isTestnet()) initContractsCITEST();
        else initContractsCI();

        vm.stopBroadcast();

        logRegisteredContracts();

        if (!__DEPLOY_SCRIPTS_DRY_RUN) {
            string memory json = getRegisteredContractsJson();

            vm.writeFile(getDeploymentsPath(string.concat("deploy-latest.json")), json);
            vm.writeFile(getDeploymentsPath(string.concat("deploy-", vm.toString(block.timestamp), ".json")), json);
        }
    }
}
```

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation

Closes #2900 

## Solution

Add a global /execution context based on `forge` subcommand used - can be set only once per program, and it can be checked by using `isContext(ForgeContext context)` env cheatcode.
`ForgeContext` have following options:
- `TestGroup`: Test group execution context (test, coverage or snapshot).
- `Test`
- `Coverage`
- `Snapshot`
- `ScriptGroup`: Script group execution context (dry run, broadcast or resume)
- `ScriptDryRun`
- `ScriptBroadcast`
- `ScriptResume`
- `Unknown`

### CODE REVIEW

File: crates/cheatcodes/src/context.rs
Comment (mds1) [inline crates/cheatcodes/src/context.rs:12]: We probably also should have:
- `Script` which equals `ScriptDryRun || ScriptBroadcast ScriptResume` (i.e. any script mode)
- Another context that returns if you are in any test context, i.e. `Test || Coverage || Snapshot`. I'm stuck on naming here though. Perhaps `Test == StandardTest || Coverage || Snapshot`?

Comment (grandizzy) [reply crates/cheatcodes/src/context.rs:12]: ah, yep, good idea, will add `Script` and `Test` for any script / test mode rename regular test context to `StandardTest`

Comment (grandizzy) [reply crates/cheatcodes/src/context.rs:12]: actually there's no need to define new contexts, will just use existing and add `isScriptContext` and `isStandardTestContext`

Comment (grandizzy) [reply crates/cheatcodes/src/context.rs:12]: changed, please check updated PR decription

File: crates/cheatcodes/spec/src/cheatcode.rs
Comment (DaniPopes) [inline crates/cheatcodes/spec/src/cheatcode.rs:128]: I think this can just be one function that takes in an enum rather than one function per variant. Then the group is also unnecessary. CC @mattsse @klkvr

Comment (klkvr) [reply crates/cheatcodes/spec/src/cheatcode.rs:128]: you mean have one function returning enum instead of several `is*` methods? I think the motivation for this impl was discussed here https://github.com/foundry-rs/foundry/issues/2900#issuecomment-1687334918

we can probably create one function which would return something like 
```solidity
enum ForgeContext {
    Test,
    Script,
    Coverage,
    Snapshot,
    DryRun,
    Broadcast
}

struct ForgeContextData {
    ForgeContext context;
    bool isTest;
    bool isScript;
}

```

Comment (grandizzy) [reply crates/cheatcodes/spec/src/cheatcode.rs:128]: > you mean have one function returning enum instead of several `is*` methods? I think the motivation for this impl was discussed here [#2900 (comment)](https://github.com/foundry-rs/foundry/issues/2900#issuecomment-1687334918)
> 

yep, indeed, that was the motivation cc @mds1

Comment (mds1) [reply crates/cheatcodes/spec/src/cheatcode.rs:128]: Another approach would be a single `isContext(ForgeContext ctx)` method where:
```solidity
struct ForgeContext {
  ScriptDryRun,
  ScriptBroadcast,
  ScriptGroup, // equal to ScriptDryRun || ScriptBroadcast
  Coverage,
  Snapshot,
  Test,
  TestGroup, // equal to Coverage || Snapshot || Test
  // etc...
}
```

Ultimately these all feel similarly clean and usable so I'm fairly indifferent. I do think the nested `ForgeContextData` struct above is my least favorite since it's a little clunkier

File: crates/forge/bin/main.rs
Comment (mds1) [inline crates/forge/bin/main.rs:127]: Should we exit with an error here and instruct the user to open an issue in this case? Ideally we should never hit this branch since all contexts should be enumerated

Comment (grandizzy) [reply crates/forge/bin/main.rs:127]: I don't think we should, since it is matched on forge subcommand then all the other subcomands like doc, debug, etc. should be handled then

File: crates/cheatcodes/src/env.rs
Comment (DaniPopes) [inline crates/cheatcodes/src/env.rs:11]: use `std::sync::OnceLock`

File: crates/cheatcodes/src/env.rs
Comment (DaniPopes) [inline crates/cheatcodes/src/env.rs:298]: derive `PartialEq, Eq` and remove discriminant calls

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/cheatcodes/src/env.rs
Comment (grandizzy) [reply crates/cheatcodes/src/env.rs]: updated in https://github.com/foundry-rs/foundry/pull/7377/commits/d3d661cca7cc34075d31a74023b8e34c248c0c72

File: crates/cheatcodes/src/env.rs
Comment (grandizzy) [reply crates/cheatcodes/src/env.rs]: updated in https://github.com/foundry-rs/foundry/pull/7377/commits/d3d661cca7cc34075d31a74023b8e34c248c0c72

File: crates/forge/tests/cli/context.rs
Comment (DaniPopes) [inline crates/forge/tests/cli/context.rs:76]: These tests don't have to be async

File: crates/cheatcodes/src/env.rs
Comment (DaniPopes) [inline crates/cheatcodes/src/env.rs:297]: This shouldn't panic if it's not set

File: crates/cheatcodes/src/env.rs
Comment (DaniPopes) [inline crates/cheatcodes/src/env.rs:296]:
```suggestion
fn is_forge_context(context: ForgeContext) -> bool {
```

**Review by DaniPopes — APPROVED** [review]
lgtm, pending another review. CC @mds1 @mattsse @klkvr

File: crates/cheatcodes/src/env.rs
Comment (DaniPopes) [inline crates/cheatcodes/src/env.rs:246]: I would prefer we call 
`let Some(current_context) = FORGE_CONTEXT.get() else { return Ok(false.abi_encode()) }`
at the top of this function and then just comparing, rather than the `is_forge_context` function.

File: crates/forge/tests/cli/context.rs
Comment (grandizzy) [reply crates/forge/tests/cli/context.rs]: yep, changed with https://github.com/foundry-rs/foundry/pull/7377/commits/4921f557ddd606ef30c37e7ceed5d8f18d0fec27

**Review by Evalir — APPROVED** [review]
lgtm! test failures seem spurious, so waiting on another review

**Comment by DaniPopes** [comment]
It's not spurious, you have to run the command locally as the output and doc suggest

**Review by mattsse — APPROVED** [review]
lgtm

**Comment by mattsse** [comment]
@grandizzy I ran `cargo cheats`, which perhaps broke a few things in sol files

**Comment by grandizzy** [comment]
> @grandizzy I ran `cargo cheats`, which perhaps broke a few things in sol files

@mattsse yeah, sorry, my bad, was something I missed for adding the new enum, should be fixed now (forced pushed your commit). @Evalir 's comment re CI failing was a legit concern :man_facepalming:

**Review by Evalir — APPROVED** [review]

File: crates/cheatcodes/spec/src/vm.rs
Comment (DaniPopes) [inline crates/cheatcodes/spec/src/vm.rs:86]: Just realized that this does not allow future expansion as Unknown will have to move, maybe move this as the first variant? @grandizzy @mattsse @Evalir

**Pull request merged**
